### ISSUE TITLE
Skin Shortcuts 1.0.0

### ISSUE THREAD

**Issue by ghost** [BigNoid/script.skinshortcuts#22]
I had a list of tasks that needed to be done within Skin Shortcuts before I was happy to declare it a 1.0.0 release and, with Isengard around the corner, I thought it would be interesting to post it here - with a few notes - to see if the script could be got up to a 1.0.0 release to coincide with Kodi J*, whenever that lands. And - more importantly - to see what others think would be necessary for a 1.0.0 release, and where they disagree with my views :)
1. Most important, this script should work for all users without crashes. That doesn't mean there won't be errors (preferable, but lets be realistic!), but the script itself should work in normal circumstances (exceptional circumstances being a broken Kodi installation, user/skinner error, etc)
   - When I last engaged fully with the community there was still the occasional issue saving the menu. I would expect PR #21 to go someway towards tracking these down, if they still happen.
2. Remove legacy code / Other code cleanup. The script in its current form is nothing like the script I originally envisioned (to be brief, it wasn't originally intended to provide the whole main menu, just - well - a list of shortcuts). The fact that this didn't appear to be what the community wanted means the code developed as necessary - and as my skills (even more limited then than they are today) allowed - to support this and features I deemed necessary to do this + requested features were added ad hoc, rather than in a planned way. It would be very nice to remove a lot of the legacy code, which would clean up the code and improve performance - but would likely break some skins if the cleanup isn't done more carefully than I could probably manage!
   - Removing .shortcuts compatibility would break Nox5 + variants, and possibly others, but would be great.
   - Menu building code has become ridiculously overloaded and needs a revamp. This would almost certainly speed up menu building and simplify code, but has the potential to break skins if not done very carefully!
   - Ideally, store main menu/submenu/additional menus in a single file. This would go someway to getting rid of the horror that is the labelID system, but stands a very good chance of breaking - and breaking badly - if not done perfectly, and is a lot of work as it affects most every area of the script. And, of course, for a 1.0.0 release, it'd be great to drop support for the then-deprecated .DATA.xml files, too...
   - Best way may be to break backwards compatibility (supporting J\* would mean 4 versions of Kodi supported in one script version - which already causes some issues regarding which shortcuts are available on which version)...
3. Feature freeze. It actually doesn't look like a huge number of features have been added since BigNoid took over day-to-day maintenance of the script (though there are some good ones that have, and huge kudos BigNoid, Phil65, marcelveldt, bambi73, et al - it's great to see the script didn't just die). But for a 1.0 release, a feature freeze - possibly mirroring Kodi J*'s - would be very useful for ensuring a stable release.
   - No potential issues?
4. Missing features. I'm sorted myself here with the templates - a feature I'd long planned and is good to finally have in core - but the most amazing thing about this script to me has always been the incredibly inventive ways skinners have utilised it. It would be nice to ensure that any 1.0.0 release includes support - minimal if necessary - for any currently identified missing features.
   - Identify missing features.

Would love to hear any comments on this, and would love even more for the script to reach a genuine 1.0.0 status :)

**Comment by marcelveldt** [comment]
1) I still see/hear an occasional report of the build failed but I believe that is caused when the users chooses to redirect the addon_data folder from Kodi to a smb location. The script can't handle not being writing to local storage (as of most python scripts). I think it's nothing to worry about.

2) I'm pretty sure that the only skin that still uses the legacy method of skinshortcuts is Aeon Nox. If needed I can help @BigNoid to migrate to the new approach in his skin without data loss.That would mean you can drop all that legacy code (which would be great because some stuff in the code is very hard to follow for me atm).

The speed is indeed a problem. When you have lot's of shortcut-overrides and conditions in the overrides.xml it can take ages for the menu to build. In my skin I just removed all default backgrounds, icons and widgets and that speeded up the menu build by factor 5 or something.

3) Great suggestion. Just focus on stability, cleaning up code and speed of the script to get the 1.0.0 release there. OR: Have that feature freeze now, fix any outstanding important bugs and do the release. After that start working on the code revamp.

4) I didn't really look into the templates yet. It looks a lot like the new feature that was added in Isengard to pass arguments to includes. Sure that this is a skinshortcuts feature that cannot be handled by the new feature in Isengard ? As for features. The script already does a great job to build menu listings. Besides the widget stuff I cannot think about any huge features. Only some small stuff to polish things up:
- condition parameter for backgrounds and icons
- change action changes (as discussed before)

These are probably thinks I'll look into next week or so...

**Comment by ghost** [comment]
1) I've always assumed they were unicode issues, but still. I'm happy with errors (there's always going to be some), not crashes - which that was doing until 21 landed.

2/3) It wouldn't just be Nox5 + variants that would be affected if all the legacy code was removed, though. There's a lot more than just the .shortcuts that has the potential to be removed. If there were to be a point were it were to be removed it makes sense to me to do it around the same time as - say - a beta release of Kodi. Chances are skinners would be starting to update skins around then, so break anything when skinners are likely to be active to fix.

Any areas that are particularly difficult to follow, feel free to flag. Even if they won't necessarily get a rewrite (and they might!), chances are that any code commenting can be improved to make it easier to follow.

The big speed issue is that, due to the way the build code has been extended over time - each menu item is actually built multiple times, along with any overrides or what not being performed multiple times. The big re-write there would be to build each menu item just once, then add visibility conditions / anything else that actually changes between the different versions that are built. Realistically, build times should be improved by anything between 50% and 65%. This must happen at some point!

4) The original idea behind templates was to add a way that submenu's can be animated between - with a traditional implementation they're all in one list, with the specific items having a visibility condition. With templates they're all in their own list, with the list itself becoming visible/hidden - so the list can be animated. Anything else in there is pure soup, and - though I still haven't fully got my head around $PARAMs - may well be handled better by Kodi directly.

**Comment by BigNoid** [comment]
I was not aware I am using a deprecated method. I removed the `.shortcuts` files now and moved to the `.DATA.xml` format. See https://github.com/BigNoid/Aeon-Nox/commit/b17f484611f247713f149dda2766eca7a39bdd09

**Comment by ghost** [comment]
Update on my ideas for what would constitute a 1.0 release:-
1. Fixes for when menu save fails (#21), along with various unicode fixes (from @roee88 - #33) should fix most known script crashes. Personally not currently aware of any other scenarios where script crashes, though obviously they should be fixed before 1.0.
2. #30 removes some legacy code. Plenty more to be removed, but now I know how to search repo for skins using legacy methods (and therefore know how to let them know to update to newer methods) this is possible...
   (and #29 + #34 are resulting in 90%+ improvement in menu build speed)
3. Feature freeze for J\* still sounds good to me ;)
4. My originally identified missing feature are now integrated, and new widget code will land soon. Condition params for background/icons - @marcelveldt - still to land. Still useful to identify any outstanding missing features.

**Comment by ghost** [comment]
Incomplete list of code that could be removed if all skins were updated to newer methods:
- Controls 101/102/103/111 - though if skinners still find this old method of showing/selecting shortcuts useful, no point in removing (though pretty much all new skins use newer methods...)
- Skin-recommended shortcuts - can now be provided by modifying overrides.xml groupings
- Widget button 309 (when 312 lands) - probably too soon to remove, though by providing 312 + custom widget-groupings, can be removed. Best to deprecate for now ;)
- Buttons 406 onwards - can now be done by setting a window property before calling 405
- (General update) Change gui to load all shortcuts on load - would provide far more ways to display/edit shortcuts, but stands a good chance of breaking skins.

I think that's all the big ones. There's numerous smaller changes that could be made which wouldn't affect (m)any skins, but would need care to check what exactly is affected before pushing.

**Comment by ghost** [comment]
Update with 0.5.3 hopefully to hit official repo soon:

GUI 309 (original widget selection) now deprecated within code (and with logging to specify so). Will be interesting to see how quickly skins that use the script to manage the whole menu transition to the new methods ;)

Biggest issue with deprecating old code (and therefore being able to fully removing said code) ahead of a 1.0.0 release remains skins which use the script in a more limited way (as the script was actually initially intended) such as Nox 5 & variants, Mimic, etc. Would be probably be interesting to actually provide PRs to these skins to see whether the skinners actually have interest in using newer methods provided, or whether the original - and simpler - methods are still actually preferred by skinners.

Also of interest of ahead of a 1.0.0 release is ensuring that any release remains simple to implement initially - with additional features being optional - as regards comments made by @Jayz2K + KOver skin on the forum.

**Comment by ghost** [comment]
Update with 0.6.6 about to hit official repo:

Aside from one major bug which was, thankfully, found before the script hit the official repo, the script continues to be remarkably stable. As such, and with Jarvis fast approaching, I'd like to suggest that the next update to the script becomes the official 1.0.0 with the following changes:-
- Jarvis changes to available shortcuts - new event logging window; changes to music links with removal of files as a separate window - potentially with a bump so new version is Jarvis only (otherwise, version equivalency functions can handle this)
- General stability improvements - adjust code so no need to check if user/script/skin overrides.xml is None (as recent issue) - so they always return an ElementTree object whether the file could be loaded or not; refactoring of all library.py functions which load a specific group of available shortcuts into a single master function (singly handle all errors, significantly reduce code therefore less chance of errors)
- Return to using json to enable/disable Kodi debug logging - see #101 

Though there are still areas of the code that could be deprecated/removed, IMHO this should allow a fully stable 1.0.0 release, where the other areas can be looked at later.

**Comment by ghost** [comment]
@marcelveldt With Jarvis around the corner, a repo push is probably required sooner rather than later with the Jarvis changes. I'm happy that most of the original intent of wanting a 1.0.0 release (except perhaps removing legacy code - which has proved almost impossible without breaking skins!) has been achieved. If we were to bump the version to 1.0.0 after #110 is merged, would you feel confident in a repo push?

**Comment by marcelveldt** [comment]
Good idea about the repo push. I was always going to do so for skin helper.
Let's wait for that final merge to happen, test a few days on the beta repo and do the update to the stable repo.

Also 100% agree with the 1.0.0 version ;-)

**Comment by ghost** [comment]
OK, I'll do a version bump PR to be merged when the repo push happens. Any thoughts on also including #105? (I know there hasn't been any feedback on the PR itself, but it's a oft-requested feature and, once its in, leaves the feature requests I'm aware of as a very short list...)

(Sounds daft, but I'm quite excited at - what is essentially an arbitrary - version bump to 1.0.0. I've worked on a fair number of projects over the years, but I can only think of one where I had any say in the version numbering ever reaching v1 before. It always feels like a big thing to do, as it feels like saying that you're happy with the state of it. In that respect, I'm a traditional programmer: to me there's always things that could be better!)

**Comment by marcelveldt** [comment]
Maybe indeed wait for that too.

**Comment by marcelveldt** [comment]
Version 1.0.0 has been submitted to the repo :-)

**Comment by ghost** [comment]
Cheers Marcel. And with that, and with great pleasure, I close this issue :)
